lib/src/native/realm_core.dart

```diff
@@ -1104,6 +1104,82 @@ class _RealmCore {
           "Switch user failed");
     });
   }
+
+  String userGetCustomData(User user) {
+    final customDataPtr = _realmLib.invokeGetPointer(() => _realmLib.realm_user_get_custom_data(user.handle._pointer));
+    try {
+      final customData = customDataPtr.cast<Utf8>().toDartString();
+      return customData;
+    } finally {
+      _realmLib.realm_free(customDataPtr.cast());
+    }
+  }
+
+  Future<void> userRefreshCustomData(App app, User user) {
+    final completer = Completer<void>();
+    _realmLib.invokeGetBool(
+        () => _realmLib.realm_app_refresh_custom_data(
+              app.handle._pointer,
+              user.handle._pointer,
+              Pointer.fromFunction(void_completion_callback),
+              completer.toPersistentHandle(),
+              _deletePersistentHandleFuncPtr,
+            ),
+        "Refresh custom data failed");
+    return completer.future;
+  }
+
+  Future<UserHandle> userLinkCredentials(App app, User user, Credentials credentials) {
+    final completer = Completer<UserHandle>();
+    _realmLib.invokeGetBool(
+        () => _realmLib.realm_app_link_user(
+              app.handle._pointer,
+              user.handle._pointer,
+              credentials.handle._pointer,
+              Pointer.fromFunction(_app_user_completion_callback),
+              completer.toPersistentHandle(),
+              _deletePersistentHandleFuncPtr,
+            ),
+        "Link credentials failed");
+    return completer.future;
+  }
+
+  int userGetState(User user) {
+    return _realmLib.realm_user_get_state(user.handle._pointer);
+  }
+
+  UserIdentity userGetIdentity(User user) {
+    final idPtr = _realmLib.invokeGetPointer(() => _realmLib.realm_user_get_identity(user.handle._pointer), "Error while getting user identity");
+    final userId = idPtr.cast<Utf8>().toDartString();
+
+    final authProvider = _realmLib.realm_user_get_auth_provider(user.handle._pointer);
+    return UserIdentityInternal.create(userId, AuthProviderType.values.fromIndex(authProvider));
+  }
+
+  List<UserIdentity> userGetIdentities(User user) {
+    return using((arena) {
+      final idsCount = arena<IntPtr>();
+      _realmLib.invokeGetBool(
+          () => _realmLib.realm_user_get_all_identities(user.handle._pointer, nullptr, 0, idsCount), "Error while getting user identities count");
+
+      final idsPtr = arena<realm_user_identity_t>(idsCount.value);
+      _realmLib.invokeGetBool(
+          () => _realmLib.realm_user_get_all_identities(user.handle._pointer, idsPtr, idsCount.value, idsCount), "Error while getting user identities");
```

Comment:
This approach is prone to race conditions due to https://github.com/realm/realm-core/issues/5430. Nothing we can do about it until the Core bug has been fixed, but we can at least leave a comment to make sure we don't forget about it.

Reply:
indeed. We can't do much about it. I will add a TODO comment